### PR TITLE
NetBSD doesn't ship with <machine/npx.h>.

### PR DESCRIPTION
**Pull request by krytarowski**
This is first step towards stopping to assume that `HAVE_BSD_REGS_T` implies
FreeBSD-style code. This breaks on NetBSD in several places.

Perhaps `HAVE_BSD_REGS_T` should be renamed to `HAVE_FREEBSD_REGS_T`.

### CODE REVIEW

**Comment by krytarowski** [comment]
@jasonwilliams200OK do you know who to ask for FreeBSD? It seems this patch failed for this system.

**Comment by jasonwilliams200OK** [comment]
@krytarowski, apparently it looks fine. So I booted FreeBSD and experimented the code in a separate directory with test.cmake file.

What you have done is a correct approach. If you look at the CMakeFiles/CMakeError.log, you will find the issue that `<machine/npx.h>` and `<machine/regs.h>` both require an additional header.

To fix this using `CheckIncludeFiles`, here is the solution:

```cmake
check_include_files("sys/types.h;machine/npx.h" HAVE_MACHINE_NPX_H)
check_include_files("sys/cdefs.h;machine/reg.h" HAVE_MACHINE_REG_H)
````

Note that this is the reason of using CheckIncludeFiles macro, instead of (singular) CheckIncludeFile, that you can pass a list of dependent headers. The list can also be set as `set(MY_HEADERS_SET sys/types.h sys/cdefs.h sys/foobar.h)` and then pass as `check_include_files(MY_HEADERS_SET variableName)`

So far so good. When succeed, it will show the message:

> Looking for include files sys/types.h, machine/npx.h - found

Minor:
However, in case of Linux, where this test is meant to fail, we will see a message `sys/types.h` and `sys/cdefs.h` not found instead of `machine/npx.h` and `machine.reg.h`. To remedy it, you can use:

```cmake
set(CMAKE_REQUIRED_QUITE 1) # muting message
check_include_files("sys/types.h;machine/npx.h" HAVE_MACHINE_NPX_H)
check_include_files("sys/cdefs.h;machine/reg.h" HAVE_MACHINE_REG_H)
set(CMAKE_REQUIRED_QUITE) # unmuting message
message(STATUS "Looking for include files sys/types.h, machine/npx.h")
if (NOT HAVE_MACHINE_NPX_H)
    message(STATUS "Looking for include files sys/types.h, machine/npx.h - not found)
else ()
    message(STATUS "Looking for include files sys/types.h, machine/npx.h - found)
endif()
message(STATUS "Looking for include files sys/cdefs.h, machine/reg.h")
if (NOT HAVE_MACHINE_REG_H)
    message(STATUS "Looking for include files sys/cdefs.h, machine/reg.h - not found)
else ()
    message(STATUS "Looking for include files sys/cdefs.h, machine/reg.h - found)
endif()
```

File: src/pal/src/thread/context.cpp [hunk 1]
```diff
@@ -50,10 +50,13 @@ extern void CONTEXT_CaptureContext(LPCONTEXT lpContext);
...
 #if !HAVE_MACH_EXCEPTIONS
 
-#if HAVE_BSD_REGS_T
+#if HAVE_MACHINE_REG_H
 #include <machine/reg.h>
+#endif
+#if HAVE_MACHINE_NPX_H
 #include <machine/npx.h>
-#endif  // HAVE_BSD_REGS_T
+#endif
```
Comment (jasonwilliams200OK) [inline src/pal/src/thread/context.cpp]: nit: add a scope end comment: // HAVE_MACHINE_NPX_H.

File: src/pal/src/thread/context.cpp [hunk 2]
```diff
@@ -50,10 +50,13 @@ extern void CONTEXT_CaptureContext(LPCONTEXT lpContext);
 
 #if !HAVE_MACH_EXCEPTIONS
 
-#if HAVE_BSD_REGS_T
+#if HAVE_MACHINE_REG_H
 #include <machine/reg.h>
+#endif
```
Comment (jasonwilliams200OK) [inline src/pal/src/thread/context.cpp]: nit: add a scope end comment: // HAVE_MACHINE_REG_H.

**Comment by krytarowski** [comment]
Thank you @jasonwilliams200OK 

How about using `CMAKE_REQUIRED_INCLUDES` instead does it work for you?

**Comment by krytarowski** [comment]
Ah, not. `CMAKE_REQUIRED_INCLUDES` is for defining include directories.

**Comment by jasonwilliams200OK** [comment]
@krytarowski, yup there is one "hammer" solution for everything: cmake_compile_c_source, which is semantically less intuitive and poorly describes the intent. So we could do:

```cmake
check_c_source_compiles("
#include <sys/cdefs.h>
#include <machine/reg.h>
int main(void){return 0;}" HAVE_MACHINE_REG_H)
```

I would try *everything* before falling back to this approach, i.e. the last resort. :smile:

**Comment by krytarowski** [comment]
Added new revision with added text:

```
    Check headers with this fashion:
    - FreeBSD requires <sys/types.h> for <machine/npx.h>
    - FreeBSD requires <sys/cdefs.h> for <machine/reg.h>
    - NetBSD doesn't need extra includes for <machine/reg.h>
    - NetBSD requires <sys/param.h> before inclusion of <sys/cdefs.h>
```

**Comment by krytarowski** [comment]
I decided to go for default CMake messages.

**Comment by jasonwilliams200OK** [comment]
SGTM. :+1:

I thnk when DragonFly and/or OpenBSD folks will chime in, then it would settle down the BSD macros once 'n forever. I discussed someone with #FreeBSD over at IRC regarding ARM support, people might participate once Linux ARM support flourishes. But until then these definitions will probably remain untouched.

**Comment by krytarowski** [comment]
Thank you for your feedback @jasonwilliams200OK

**Pull request merged**